xmrstak/backend/amd/autoAdjust.hpp

```diff
@@ -97,34 +95,46 @@ class autoAdjust
 		int i = 0;
 		for(auto& ctx : devVec)
 		{
-			/* 1000 is a magic selected limit, the reason is that more than 2GiB memory
-			 * sowing down the memory performance because of TLB cache misses
-			 */
-			size_t maxThreads = 1000u;
-			if(ctx.name.compare("gfx901") == 0)
-			{
-				/* Increase the number of threads for AMD VEGA gpus.
-				 * Limit the number of threads based on the issue: https://github.com/fireice-uk/xmr-stak/issues/5#issuecomment-339425089
-				 * to avoid out of memory errors
-				 */
-				maxThreads = 2024u;
-			}
-
-			// keep 128MiB memory free (value is randomly chosen)
-			size_t availableMem = ctx.freeMem - (128u * byteToMiB);
-			// 224byte extra memory is used per thread for meta data
+			/*
+			  GCN has a hard architectural limit of 64 CUs max (Invalid for other archs though).
+			  Reserve 1 CU for the system. 
+			*/
+			size_t maxComputeUnitsAvailable = std::max(std::min(ctx.computeUnits, 64) - 1, 1);
+			//Each CU is 64 threads executed in an SIMT fashion:
+			size_t maxThreadsAvailable = maxComputeUnitsAvailable << 6;
+			//Keep 128MiB memory free (value is randomly chosen):
+			size_t availableMem = ctx.freeMem - (128u << 20);
+			//224byte extra memory is used per thread for meta data
 			size_t perThread = hashMemSize + 224u;
 			size_t maxIntensity = availableMem / perThread;
-			size_t possibleIntensity = std::min( maxThreads , maxIntensity );
-			// map intensity to a multiple of the compute unit count, 8 is the number of threads per work group
-			size_t intensity = (possibleIntensity / (8 * ctx.computeUnits)) * ctx.computeUnits * 8;
-			conf += std::string("  // gpu: ") + ctx.name + " memory:" + std::to_string(availableMem / byteToMiB) + "\n";
+			size_t possibleIntensity = std::min(maxThreadsAvailable, maxIntensity);
+			//Alias intensity against the smallest work group size possible:
+			size_t intensity = (possibleIntensity >> 3) << 3;
+			conf += std::string("  // gpu: ") + ctx.name + " memory:" + std::to_string(availableMem >> 20) + "\n";
 			conf += std::string("  // compute units: ") + std::to_string(ctx.computeUnits) + "\n";
-			// set 8 threads per block (this is a good value for the most gpus)
-			conf += std::string("  { \"index\" : ") + std::to_string(ctx.deviceIdx) + ",\n" +
+			//set 8 threads per block (this is a good value for the most gpus)
+			//Create two instances if more than one CU is available:
+			if (maxComputeUnitsAvailable > 1)
+			{
+				//Reports seem to indicate splitting a physical device into two works better:
+				conf += std::string("  { \"index\" : ") + std::to_string(ctx.deviceIdx) + ",\n" +
```

Comment:
please move the config generation after the if condition and collect all needed states into helper variables.